src/bin/derper.rs

```diff
@@ -0,0 +1,846 @@
+        tokio::fs::write(path, config_ser)
+            .await
+            .context("unable to write config file")?;
+
+        Ok(())
+    }
+}
+
+const DEV_PORT: u16 = 3340;
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    tracing_subscriber::registry()
+        .with(tracing_subscriber::fmt::layer().with_writer(std::io::stderr))
+        .with(EnvFilter::from_default_env())
+        .init();
+
+    let mut cli = Cli::parse();
+
+    let mut tasks = JoinSet::new();
+
+    if cli.dev {
+        cli.addr = SocketAddr::new(IpAddr::V4(Ipv4Addr::UNSPECIFIED), DEV_PORT);
+        info!(%cli.addr, "Running in dev mode.");
+    }
+
+    let listen_host = cli.addr.ip();
+    let serve_tls = cli.addr.port() == 443 || CertMode::Manual == cli.cert_mode;
+
+    let derp_server = if cli.run_derp {
+        let cfg = Config::load(&cli).await?;
+
+        let mesh_key = if let Some(file) = cli.mesh_psk_file {
+            let raw = tokio::fs::read_to_string(file)
+                .await
+                .context("reading mesh-pks file")?;
+            let mut mesh_key = [0u8; 32];
+            hex::decode_to_slice(raw.trim(), &mut mesh_key).context("invalid mesh-pks content")?;
+            info!("DERP mesh key configured");
+            Some(mesh_key)
+        } else {
+            None
+        };
+        let derp_server: derp::Server<OwnedReadHalf, OwnedWriteHalf, derp::HttpClient> =
+            derp::Server::new(cfg.private_key, mesh_key);
+        info!("DERP server configured");
+        Some(derp_server)
+    } else {
+        None
+    };
+
+    if cli.run_stun {
+        tasks.spawn(async move { serve_stun(listen_host, cli.stun_port).await });
+    }
+
+    let client_conn_handler = derp_server.map(|s| {
+        let headers = if serve_tls {
+            HeaderMap::from_iter(
+                TLS_HEADERS
+                    .iter()
+                    .map(|(k, v)| (k.parse().unwrap(), v.parse().unwrap())),
+            )
+        } else {
+            Default::default()
+        };
+        s.client_conn_handler(headers)
+    });
+    let tls_config = if serve_tls {
+        let contact = "d@iroh.computer".to_string(); // TODO: configurable.
+        let is_production = false; // TODO: configurable
+        let (config, acceptor) = cli
+            .cert_mode
+            .gen_server_config(
+                cli.hostname.clone(),
+                contact,
+                is_production,
+                cli.cert_dir.unwrap_or_else(|| PathBuf::from(".")),
+            )
+            .await?;
+        Some((config, acceptor))
+    } else {
+        None
+    };
+
+    let server = Derper {
+        client_conn_handler,
+        tls_config,
+    };
+
+    server.run(cli.addr, cli.http_port).await?;
+
+    // Shutdown all tasks
+    tasks.abort_all();
+
+    Ok(())
+}
+
+const NO_CONTENT_CHALLENGE_HEADER: &str = "X-Tailscale-Challenge";
+const NO_CONTENT_RESPONSE_HEADER: &str = "X-Tailscale-Response";
```

Comment:
doesn't look like this is terribly important, but we should probably change these headers.